hibernate-core/src/main/java/org/hibernate/proxy/pojo/javassist/JavassistProxyFactory.java

```diff
@@ -93,7 +93,21 @@ public void postInstantiate(
 		javassist.util.proxy.ProxyFactory factory = new javassist.util.proxy.ProxyFactory() {
 			@Override
 			protected ClassLoader getClassLoader() {
-				return persistentClass.getClassLoader();
+				// This classloader does not take into account class space consistency
+				// i.e. if the persistenceClass can see the same interface types via
+				// a different ClassLoader hierarchy. In that case we could try some 
+				// risky magic with loadClass, but using findClass is much safer.
+				return new ClassLoader(persistentClass.getClassLoader()) {
```

Comment:
@brmeyer And I am a little shaky on what exactly (Class loading wise) Javassist should be given access to here :)
I need to understand that better to give you a better answer.

I think creating the new ClassLoader each and every call is going to be bad here.  Javassist uses this as a cache key; we might get lucky there in that we really do not need the "ProxyFactory caching" based on how we use Javassist.  I am missing the logic why a new ClassLoader each time is being proposed.

We may also want to look at using `ProxyFactory#classLoaderProvider` rather than `ProxyFactory#getClassLoader`.  The difference is `#classLoaderProvider` is set once, whereas `#getClassLoader` is overridden each time.  IIRC we avoided `ProxyFactory#classLoaderProvider` because it was added later and we wanted to maintain some level of back-compat.